feat: Add Support for files at runtime

Warlord-K opened the pull request:
This PR adds support for files to be sent to cloud to be available at runtime with `fal run` and `fal deploy`. The interface will allow users to add relative and absolute paths of files and directories to be available at runtime. 

```python
class MyApp(
    App, 
):
    files = [
        "text_file.txt", 
        "python_file.py",
        "somefolder",
        "/Users/user/dir/folder",
    ]

    def setup(self):
        # Python imports available
        from python_file import ...

        # Files and folders available at relative path in /deploy-data
        with open("/deploy-data/folder/records.json") as f:
            contents = f.read()
```

We also have a local cache at `~/.cache/files_metadata.json` which makes subsequent runs and deploys ultrafast.

This PR is still a WIP, adding tests and better error messages when files aren't found etc.

Warlord-K commented on projects/fal/pyproject.toml:
```diff
@@ -23,7 +23,7 @@ readme = "README.md"
 dependencies = [
     "isolate[build]>=0.18.0,<0.19.0",
-    "isolate-proto>=0.12.0,<0.13.0",
+    "isolate_proto @ git+https://github.com/fal-ai/fal.git@isolate-proto-files#subdirectory=projects/isolate_proto",
```
I will update this when #604 gets merged

chamini2 commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,341 @@
+        sha256.update(metadata_string.encode("utf-8"))
+
+        return sha256.hexdigest()
+
+    def get_file_metadata(self, file_path: str) -> dict[str, Any]:
+        stat = os.stat(file_path)
+        file_hash = self.compute_hash(file_path, stat.st_mtime, stat.st_mode)
+        return {
+            "size": str(stat.st_size),
+            "mtime": str(stat.st_mtime),
+            "mode": str(stat.st_mode),
+            "hash": file_hash,
+        }
```
Also, why do we include st info at all in the hash instead of just comparing hash from the file and also copmaring manually the mode?

Warlord-K replied:
This is to ensure that the cloud can distinguishes between files with different mode. The idea was to create a single hash which can be used to check file existence on cloud, otherwise when the mode changes we would still be using the old file or have to add additional checks everywhere for the file mode

Warlord-K commented on projects/fal/src/fal/api.py:
```diff
@@ -422,6 +426,9 @@ class FalServerlessHost(Host):
     _thread_pool: ThreadPoolExecutor = field(default_factory=ThreadPoolExecutor)
 
+    def set_local_file_path(self, path: str):
+        self.local_file_path = path
```
Alright

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,341 @@
+
+        creds = get_default_credentials()
+        return client.TusClient(
+            f"{REST_URL}/files/tus",
```
Tus is the route for the file uploads, TUS is the protocol I am using for resumable uploads. Its defined here, https://tus.io 

I used it since people might want to upload large files and then having the ability to resume in case of errors would be helpful

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,341 @@
+        sha256 = hashlib.sha256()
+
+        # Hash file contents
+        with open(file_path, "rb") as f:
+            for chunk in iter(lambda: f.read(4096), b""):
+                sha256.update(chunk)
```
I was hyperfocused on ensuring that large files are iteratively loaded in chunks and hashed but I missed that file_digest also does the same, updating

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,341 @@
+class FileSync:
+    def __init__(self, local_file_path: str):
+        self.local_file_path = local_file_path
+        self.cache_dir = Path.home() / ".cache" / "fal"
+        self.cache_dir.mkdir(parents=True, exist_ok=True)
+        self.cache_file = self.cache_dir / "file_metadata.json"
```
Got it

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,341 @@
+        self.local_file_path = local_file_path
+        self.cache_dir = Path.home() / ".cache" / "fal"
+        self.cache_dir.mkdir(parents=True, exist_ok=True)
+        self.cache_file = self.cache_dir / "file_metadata.json"
```
Each fetch was taking 0.3s earlier, which is too long for one file. As for the local cache, when there is a local cache miss we check the cloud and add to the local cache, this means even when teammates have different caches it will equalize after one run of the file.

This point did make me realize that I did not account for a person working as different users on the same machines for the cache location which is now fixed!

chamini2 replied:
OK, and why fetch per file and not just fetch once per `fal run` / `fal deploy`? This seems like over-optimizing and can lead to complications as a start.

Warlord-K replied:
Yeah! We do only fetch when we do fal run and fal deploy. The local cache idea was mainly to speed up iteration when someone is doing fal run again and again with 100s of files, with the local cache the time is nearly instant and removes any file overhead

Warlord-K replied:
Yes, now its consistently taking 1s per fal run or deploy to sync files

isidentical commented on projects/fal/src/fal/api.py:
```diff
@@ -503,6 +512,15 @@ def register(
+        files = options.host.get("files", [])
+        if files:
+            sync = FileSync(self.local_file_path)
+            result = asyncio.run(sync.sync_files(files))
```
you should never use asyncio.run inside a sync function that is not top level. either make this function async, or make sync_files work syncronusly.

isidentical commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,360 @@
+        except ValueError:
+            relative_path = absolute_path.as_posix()
+
+        # Remove parent directory traversals for security
+        while relative_path.startswith("../"):
+            relative_path = relative_path[3:]
```
why is this handled here? this feels completely useless if we are not already handling it in the backend, and then you can always use `../` inside the path so it doesn't have to start with it

Warlord-K replied:
You are absolutely correct! It is being handled in the backend too but to reduce debugging time for user, it felt simpler to check here and error out before the server gets involved

I noted your point about sanitization of the path and have improved it to be more extensive than checking leading `../`. I initially was checking all `..` but that was messing up some edge cases, I have properly fixed things so this should be better

isidentical reviewed (changes requested):
the whole API design here of local cache, hash comparisons, etc. feels inefficient. I'd recommend redesigning this in a way that you send a tree of hashes to the server, and it processes and tells you what is different which then you upload. instead of handling this logic on the client side and doing one by one comparison.

Another big thing is transacitons, we need to make sure we don't leave anything hanging and each "sync" is its own transaction and atomically renamed after the upload is complete to prevent clashes.

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,360 @@
+import asyncio
```
I do stick with pathlib most of the time but, os.path.relpath is able to correctly resolve the relative path when the relative file is outside the base directory, Implementing it myself with Pathlib could have edge cases hence I stick with os.path.relpath for the resolution. I checked and confirmed that Path.relative_to does not work when the path is outside the directory

I have taken this into account more and removed any other instances of os.path and used Pathlib wherever I could

Warlord-K commented on projects/fal/src/fal/file_sync.py:
```diff
@@ -0,0 +1,360 @@
+
+        # Upload files concurrently
+        tasks = [upload_single_file(metadata) for metadata in files_to_upload]
+        return await asyncio.gather(*tasks)
```
Definitely didn't intend that to happen, I was under the impression that the network bandwidth would be the limiting factor since I myself saw not more than 5-10 files being transferred at a time, I have added explicit bounds and will keep this in mind with async/parallel tasks!

Warlord-K commented:
> the whole API design here of local cache, hash comparisons, etc. feels inefficient. I'd recommend redesigning this in a way that you send a tree of hashes to the server, and it processes and tells you what is different which then you upload. instead of handling this logic on the client side and doing one by one comparison.

Understood! That was the initial design but upon some testing I found that users would still have access to the raw files and can remove them even though we are giving read-only access, this would mean any (fast) bulk check wouldn't be possible and came up with the local cache and everything. I have removed the local cache and changed it to a single server call with hashes to find missing ones and added some handling server side which would capture corrupted files and handle them properly so that they can be reuploaded and the single hashes check is possible!

> Another big thing is transacitons, we need to make sure we don't leave anything hanging and each "sync" is its own transaction and atomically renamed after the upload is complete to prevent clashes.

Each sync is now one transaction when the files are present, when they are not it is broken down into uploads, but these are tus uploads which are handled on the server to each be atomic, and if anything fails there the user can simply rerun and the missing files will be uploaded. Uploads are idempotent and have atomic persistence, so we should not leave anything dangling and have any clashes on files!
